codec/src/main/java/io/netty/handler/codec/MessageToBufferedByteEncoder.java

```diff
@@ -0,0 +1,302 @@
+ *                 throws {@link Exception} {
+ *             out.writeInt(msg);
+ *         }
+ *     }
+ * </pre>
+ */
+public abstract class MessageToBufferedByteEncoder<I> extends MessageToByteEncoder<I> implements ChannelInboundHandler {
+
+    private ByteBuf currentBuffer;
+    private AggregatedChannelProgressivePromise aggregatedPromise;
+
+    /**
+     * @see {@link MessageToByteEncoder#MessageToByteEncoder()}
+     */
+    protected MessageToBufferedByteEncoder() {
+        CodecUtil.ensureNotSharable(this);
+    }
+
+    /**
+     * @see {@link MessageToByteEncoder#MessageToByteEncoder(Class)}
+     */
+    protected MessageToBufferedByteEncoder(Class<? extends I> outboundMessageType) {
+        super(outboundMessageType);
+        CodecUtil.ensureNotSharable(this);
+    }
+
+    /**
+     * @see {@link MessageToByteEncoder#MessageToByteEncoder(boolean)}
+     */
+    protected MessageToBufferedByteEncoder(boolean preferDirect) {
+        super(preferDirect);
+        CodecUtil.ensureNotSharable(this);
+    }
+
+    /**
+     * @see {@link MessageToByteEncoder#MessageToByteEncoder(Class, boolean)}
+     */
+    protected MessageToBufferedByteEncoder(Class<? extends I> outboundMessageType, boolean preferDirect) {
+        super(outboundMessageType, preferDirect);
+        CodecUtil.ensureNotSharable(this);
+    }
+
+    @Override
+    public final void write(ChannelHandlerContext ctx, Object msg, final ChannelPromise promise) {
+        ByteBuf buffer = currentBuffer;
+        final ChannelOutboundBuffer outboundBuffer = ctx.channel().unsafe().outboundBuffer();
+        final int writerIndex;
+        final int readerIndex;
+        if (buffer == null) {
+            writerIndex = 0;
+            readerIndex = 0;
+        } else {
+            writerIndex = buffer.writerIndex();
+            readerIndex = buffer.readerIndex();
+        }
+        try {
+            if (acceptOutboundMessage(msg)) {
+                @SuppressWarnings("unchecked")
+                I cast = (I) msg;
+                if (buffer == null) {
+                    buffer = currentBuffer = allocateBuffer(ctx, cast, preferDirect);
+                    assert aggregatedPromise == null;
+                    aggregatedPromise = new AggregatedChannelProgressivePromise(ctx);
+                }
+                try {
+                    encode(ctx, cast, buffer);
+                } finally {
+                    ReferenceCountUtil.release(cast);
+                }
+                int readableBytes = buffer.readableBytes();
+                if (outboundBuffer != null) {
+                    int readable = readableBytes - writerIndex - readerIndex;
+                    if (readable < 0) {
+                        throw new IllegalStateException("readableBytes are less then before calling encode(...)");
+                    }
+                    outboundBuffer.incrementPendingOutboundBytes(readable);
```

Comment:
This crosses API boundaries and exposes an internal API which I would prefer to avoid. I don't necessarily have a better idea, but we may have the same issue with `SslHandler` or other handlers which aggregate writes (e.g. `ProxyHandler`).